### PR TITLE
Ensure Hints do not cause memory leaks

### PR DESCRIPTION
**Pull request by markushi**
## :scroll: Description
Clear any extra hint attributes before passing them to the transport layer.

## :bulb: Motivation and Context
Hint objects are only freed after network transfer is completed. Thus this may causes memory leaks whenever a hint contains references to an activity/fragment/view.

Options considered:
- Use `WeakReference<Object>` for hint values: probably a bad idea because calling `hint.set("key", new Value())` gives you no guarantee that  the value will be present at a later stage / could be gc'd at any time
- Use existing API, pass all relevant values as a `WeakReference` directly (e.g. `hint.set("activity", new WeakReference(activity))`) and unwrap the `WeakReference` in `get()` automatically: Would work nicely but breaks the API for anyone already using `WeakReference` as we now would suddenly auto-unwrap it.

## :green_heart: How did you test it?
Added unit test

## :pencil: Checklist

- [x] I reviewed the submitted code
- [x] I added tests to verify the changes
- [ ] I updated the docs if needed
- [x] No breaking changes

### CODE REVIEW

**Comment by markushi** [comment]
@romtsn I'm not very happy with the solution, as the presence of the activity/fragment/view hint value is not guaranteed anymore. Would it be feasible to delete the whole hint earlier and not wait until network request is done?

**Comment by adinauer** [comment]
> breaks the API for anyone already using WeakReference as we now would suddenly auto-unwrap it.

Isn't this also true for the current implementation? It doesn't matter whether it was us who set the `WeakReference` or the user. In `get()` it'll be unwrapped either way.

If we want to avoid this, I see two ways atm:
- Store a wrapper object that tells us whether it was us who used the weak ref
- Store weak refs set by us in a different map

>  presence of the activity/fragment/view hint value is not guaranteed anymore

API always said `@Nullable` but users may have tested and found they never receive null where they now will if we go ahead with this change. I guess not bloating memory is more important here.

> Would it be feasible to delete the whole hint earlier and not wait until network request is done?

Maybe we could only remove stuff we would mark as `weak` early?

**Comment by romtsn** [comment]
> @romtsn I'm not very happy with the solution, as the presence of the activity/fragment/view hint value is not guaranteed anymore. Would it be feasible to delete the whole hint earlier and not wait until network request is done?

I think we can't delete the hint entirely, because we check if it's Retryable, etc. But I think we could drop the specific keys (like activity or view) after beforeSend is executed

**Comment by adinauer** [comment]
How about we wrap stuff in the internal storage with some object that allows us to mark things as weak without actually using `WeakReference` and then have a method on `Hint` to remove weak stuff after `beforeSend` was executed?

How does this work for things that are not errors? Do we also attach things that should be weakly referenced there? Then we would have to make sure everything is cleaned up which is more risky than using `WeakReference`

**Comment by romtsn** [comment]
> How does this work for things that are not errors (e.g. transactions)? Do we also attach things that should be weakly referenced there? Then we would have to make sure everything is cleaned up which is more risky than using WeakReference

Yes, we also send breadcrumbs with transactions, so they can leak stuff. I guess we can just drop the keys inside `transport.send`, because that's called always, regardless of the envelope type

**Comment by adinauer** [comment]
I assume hints for `beforeBreadcrumb` are fine because we don't store them. Seems like the only problematic hints are those passed to the transport so clearing in transport sounds like a good approach to me.

**Comment by markushi** [comment]
After having a closer look with @romtsn the issue is less relevant as initially thought:
* Affected Hints for breadcrumbs are processed right after adding the hint and are not passed to the transportation layer
* For events: the transport layer does not read any hint attributes, thus those could be safely cleared to avoid any leak

Let me adopt the PR accordingly

**Review by adinauer — APPROVED** [review]

File: CHANGELOG.md [hunk 1]
```diff
@@ -7,6 +7,7 @@
 - Use `canonicalName` in Fragment Integration for better de-obfuscation ([#2379](https://github.com/getsentry/sentry-java/pull/2379))
 - Fix Timber and Fragment integrations auto-installation for obfuscated builds ([#2379](https://github.com/getsentry/sentry-java/pull/2379))
 - Don't attach screenshots to events from Hybrid SDKs ([#2360](https://github.com/getsentry/sentry-java/pull/2360))
+- Fix ensure Hints do not cause memory leaks ([#2387](https://github.com/getsentry/sentry-java/pull/2387))
```
[the remark this answers is not in the document]
Comment (adinauer) [reply CHANGELOG.md:10]: `l`

File: sentry/src/main/java/io/sentry/SentryClient.java [hunk 2]
```diff
@@ -179,6 +179,7 @@ private boolean shouldApplyScopeData(
           buildEnvelope(event, attachments, session, traceContext, null);
 
       if (envelope != null) {
+        hint.clear();
```
Comment (romtsn) [inline sentry/src/main/java/io/sentry/SentryClient.java:182]: I think we can still clear it, even if the envelope is null, so probably just move it before `if`

File: sentry/src/main/java/io/sentry/SentryClient.java [hunk 3]
```diff
@@ -577,6 +579,7 @@ public void captureSession(final @NotNull Session session, final @Nullable Hint
               transaction, filterForTransaction(getAttachments(hint)), null, traceContext, null);
 
       if (envelope != null) {
```
Comment (romtsn) [inline sentry/src/main/java/io/sentry/SentryClient.java:581]: same here probably

**Review by romtsn — APPROVED** [review]

**Pull request merged**